From: Harald Freudenberger <freude@de.ibm.com>
Subject: kernel: Fix memory leak on EP11 target list processing.
Patch-mainline: no (SLES only)
References: bnc#1096751, LTC#168596

Description:  kernel: Fix memory leak on EP11 target list processing.
Symptom:      Kernel memory leak when EP11 target lists are used.
Problem:      freeing of the EP11 target list buffer is missing.
Solution:     free the list after use.
Reproduction: Run application which sends EP11 crypto requests and
              uses target lists (e.g. via opencryptoki and use the
              ep11.conf to limit the adapters/domains to use).
              the help of chzcrypt. Monitor top or free output.


Signed-off-by: Harald Freudenberger <freude@de.ibm.com>
Acked-by: Johannes Thumshirn <jthumshirn@suse.de>
---
 drivers/s390/crypto/zcrypt_api.c |    5 +++++
 1 file changed, 5 insertions(+)

--- a/drivers/s390/crypto/zcrypt_api.c
+++ b/drivers/s390/crypto/zcrypt_api.c
@@ -629,7 +629,10 @@ static long zcrypt_send_ep11_cprb(struct
 				   (struct ep11_target_dev __force __user *)
 				   xcrb->targets, xcrb->targets_num *
 				   sizeof(struct ep11_target_dev)))
+		{
+			kfree(ep11_dev_list.targets);
 			return -EFAULT;
+		}
 	}
 
 	spin_lock_bh(&zcrypt_device_lock);
@@ -661,9 +664,11 @@ static long zcrypt_send_ep11_cprb(struct
 		put_device(&zdev->ap_dev->device);
 		zcrypt_device_put(zdev);
 		spin_unlock_bh(&zcrypt_device_lock);
+		kfree(ep11_dev_list.targets);
 		return rc;
 	}
 	spin_unlock_bh(&zcrypt_device_lock);
+	kfree(ep11_dev_list.targets);
 	return -ENODEV;
 }
 
